storage-proofs/porep/src/nse/circuit/compound.rs

```diff
@@ -25,11 +28,129 @@ impl<'a, Tree: 'static + MerkleTreeTrait, G: 'static + Hasher>
     CompoundProof<'a, NarrowStackedExpander<'a, Tree, G>, NseCircuit<'a, Tree, G>> for NseCompound
 {
     fn generate_public_inputs(
-        pub_in: &<NarrowStackedExpander<Tree, G> as ProofScheme>::PublicInputs,
-        pub_params: &<NarrowStackedExpander<Tree, G> as ProofScheme>::PublicParams,
+        public_inputs: &<NarrowStackedExpander<Tree, G> as ProofScheme>::PublicInputs,
+        public_params: &<NarrowStackedExpander<Tree, G> as ProofScheme>::PublicParams,
         k: Option<usize>,
     ) -> Result<Vec<Fr>> {
-        todo!();
+        let config = &public_params.config;
+
+        let mut inputs = Vec::new();
+
+        // replica id
+        inputs.push(public_inputs.replica_id.into());
+
+        // comm_d
+        inputs.push(public_inputs.tau.comm_d.into());
+
+        // comm_r
+        inputs.push(public_inputs.tau.comm_r.into());
+
+        let por_setup_params = por::SetupParams {
+            leaves: config.num_nodes_sector(),
+            private: true,
+        };
+        let por_params = por::PoR::<Tree>::setup(&por_setup_params)?;
+        let por_params_d = por::PoR::<BinaryMerkleTree<G>>::setup(&por_setup_params)?;
+        let challenges = Challenges::new(
+            config,
+            public_params.num_layer_challenges,
+            &public_inputs.replica_id,
+            public_inputs.seed,
+        );
+
+        let data_inclusion_inputs =
+            |c: u64| generate_inclusion_inputs::<BinaryMerkleTree<G>>(&por_params_d, c as usize, k);
+        let layer_inclusion_inputs =
+            |c: u64| generate_inclusion_inputs::<Tree>(&por_params, c as usize, k);
+        let parent_inclusion_inputs =
+            |c: usize| generate_inclusion_inputs::<Tree>(&por_params, c as usize, k);
+
+        // layer proofs
+        for layer_challenge in challenges {
+            // -- first layer
+            {
+                let c = layer_challenge.first_layer_challenge.absolute_index;
+
+                // challenge input
+                inputs.push(u64_into_fr(c));
+
+                // comm_d inclusion proof for the data leaf
+                inputs.extend(data_inclusion_inputs(c)?);
+
+                // layer_inclusion proof
+                inputs.extend(layer_inclusion_inputs(c)?);
```

Comment:
@dignifiedquire This is somewhat redundant, right? This will result in `c` being passed three times, I think. I am not saying it is necessarily convenient to do otherwise (although it might be worth trying) — just checking to make sure I am tracking what you have here.

Reply:
While `c` is being used to generate the inputs, it is only allocated in the circuit once (the first one where we call `u64_into_fr` (as far as I remember)